src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1038,6 +1040,31 @@ void MergeTreeData::loadDataPartsFromDisk(
         {
             part->loadColumnsChecksumsIndexes(require_part_metadata, true);
         }
+#if USE_AWS_S3
+        /// This code looks really strange. Why can it happen? When we remove something from S3 we can receive different kinds of errors during
+        /// interaction with [Zoo]Keeper, S3 and even local disk. In such way we can get into situation where part is partially removed (at least we
+        /// had an intention to remove it) but server was restarted and now we are trying to load this partially removed part. It can throw an exception
```

Comment:
Don't we first rename the part to `tmp_delete_*`?
So, unsuccessful removal will only leave garbage in the filesystem.

If not, why?

Reply:
indeed, why not?